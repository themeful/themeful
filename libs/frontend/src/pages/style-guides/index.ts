export * from './style-guides.component'
