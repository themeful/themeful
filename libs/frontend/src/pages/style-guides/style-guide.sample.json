[
  {
    "name": "Global",
    "slug": "global",
    "baseFontSize": 16,
    "types": [
      {
        "name": "Color",
        "groups": [
          {
            "name": "Base",
            "styles": [
              {
                "name": "Black",
                "type": "color",
                "group": "base",
                "value": "#333333",
                "slug": "base_black"
              },
              {
                "name": "LIght",
                "type": "color",
                "group": "base",
                "value": "#ffffff",
                "slug": "base_lIght"
              }
            ]
          },
          {
            "name": "Gray",
            "styles": [
              {
                "name": "100",
                "type": "color",
                "group": "gray",
                "value": "rgba(0, 0, 0, 0.75)",
                "slug": "gray_100"
              },
              {
                "name": "200",
                "type": "color",
                "group": "gray",
                "value": "rgba(0, 0, 0, 0.60)",
                "slug": "gray_200"
              },
              {
                "name": "300",
                "type": "color",
                "group": "gray",
                "value": "rgba(0, 0, 0, 0.45)",
                "slug": "gray_300"
              },
              {
                "name": "400",
                "type": "color",
                "group": "gray",
                "value": "rgba(0, 0, 0, 0.30)",
                "slug": "gray_400"
              },
              {
                "name": "500",
                "type": "color",
                "group": "gray",
                "value": "rgba(0, 0, 0, 0.15)",
                "slug": "gray_500"
              }
            ]
          }
        ]
      },
      {
        "name": "MediaQuery",
        "groups": [
          {
            "name": "Media query",
            "styles": [
              {
                "name": "Above Small Mobile",
                "type": "mediaquery",
                "group": "mediaQuery",
                "value": "screen and (min-width: 321px)",
                "slug": "mediaQuery_aboveSmallMobile"
              },
              {
                "name": "Above Mobile",
                "type": "mediaquery",
                "group": "mediaQuery",
                "value": "screen and (min-width: 664px)",
                "slug": "mediaQuery_aboveMobile"
              },
              {
                "name": "Above Tablet",
                "type": "mediaquery",
                "group": "mediaQuery",
                "value": "screen and (min-width: 992px)",
                "slug": "mediaQuery_aboveTablet"
              },
              {
                "name": "Above Desktop",
                "type": "mediaquery",
                "group": "mediaQuery",
                "value": "screen and (min-width: 1200px)",
                "slug": "mediaQuery_aboveDesktop"
              }
            ]
          }
        ]
      },
      {
        "name": "Size",
        "groups": [
          {
            "name": "Space",
            "styles": [
              {
                "type": "size",
                "group": "space",
                "name": "Tiny",
                "value": "4px",
                "slug": "space_tiny"
              },
              {
                "type": "size",
                "group": "space",
                "name": "Small",
                "value": "8px",
                "slug": "space_small"
              },
              {
                "name": "Small Medium",
                "type": "size",
                "group": "space",
                "value": "0.75rem",
                "slug": "space_smallMedium"
              },
              {
                "type": "size",
                "group": "space",
                "name": "Medium",
                "value": "16px",
                "slug": "space_medium"
              },
              {
                "type": "size",
                "group": "space",
                "name": "Large",
                "value": "24px",
                "slug": "space_large"
              },
              {
                "type": "size",
                "group": "space",
                "name": "X-Large",
                "value": "32px",
                "slug": "space_xlarge"
              }
            ]
          }
        ]
      }
    ]
  },
  {
    "name": "StyleGuide 1",
    "slug": "styleGuide1",
    "baseFontSize": 16,
    "types": [
      {
        "name": "Font",
        "groups": [
          {
            "name": "Font",
            "styles": [
              {
                "name": "Primary",
                "type": "font",
                "group": "font",
                "value": "Verdana, Geneva, Tahoma, sans-serif",
                "slug": "font_primary"
              }
            ]
          }
        ]
      },
      {
        "name": "Color",
        "groups": [
          {
            "name": "Action",
            "styles": [
              {
                "name": "Primary",
                "type": "color",
                "group": "action",
                "value": "#31ed31",
                "slug": "action_primary"
              },
              {
                "name": "Secondary",
                "type": "color",
                "group": "action",
                "value": "#2ec22e",
                "slug": "action_secondary"
              },
              {
                "name": "Tertiary",
                "type": "color",
                "group": "action",
                "value": "#1e961e",
                "slug": "action_tertiary"
              }
            ]
          },
          {
            "name": "Brand",
            "styles": [
              {
                "type": "color",
                "group": "brand",
                "name": "Primary",
                "value": "#ff0000",
                "slug": "brand_primary"
              },
              {
                "name": "Secondary",
                "type": "color",
                "group": "brand",
                "value": "#d42828",
                "slug": "brand_secondary"
              },
              {
                "name": "Tertiary",
                "type": "color",
                "group": "brand",
                "value": "#af2323",
                "slug": "brand_tertiary"
              }
            ]
          }
        ]
      },
      {
        "name": "Font Size",
        "groups": [
          {
            "name": "Content",
            "styles": [
              {
                "type": "font-size",
                "group": "content",
                "name": "Primary",
                "value": "1.2rem",
                "slug": "content_primary"
              }
            ]
          }
        ]
      },
      {
        "name": "MediaQuery",
        "groups": [
          {
            "name": "Media query",
            "styles": [
              {
                "name": "Above Desktop",
                "type": "mediaquery",
                "group": "mediaQuery",
                "value": "screen and (min-width: 1100px)",
                "slug": "mediaQuery_aboveSmallDesktop"
              }
            ]
          }
        ]
      }
    ]
  },
  {
    "name": "StyleGuide 2",
    "slug": "styleGuide2",
    "baseFontSize": 16,
    "types": [
      {
        "name": "Font",
        "groups": [
          {
            "name": "Font",
            "styles": [
              {
                "name": "Primary",
                "type": "font",
                "group": "font",
                "value": "'Times New Roman', Times, serif",
                "slug": "font_primary"
              }
            ]
          }
        ]
      },
      {
        "name": "Color",
        "groups": [
          {
            "name": "Action",
            "styles": [
              {
                "name": "Primary",
                "type": "color",
                "group": "action",
                "value": "#ff5555",
                "slug": "action_primary"
              },
              {
                "name": "Secondary",
                "type": "color",
                "group": "action",
                "value": "#d22828",
                "slug": "action_secondary"
              },
              {
                "name": "Tertiary",
                "type": "color",
                "group": "action",
                "value": "#9f1c1c",
                "slug": "action_tertiary"
              }
            ]
          },
          {
            "name": "Brand",
            "styles": [
              {
                "name": "Primary",
                "type": "color",
                "group": "brand",
                "value": "#577bf1",
                "slug": "brand_primary"
              },
              {
                "name": "Secondary",
                "type": "color",
                "group": "brand",
                "value": "#4646ed",
                "slug": "brand_secondary"
              },
              {
                "name": "Tertiary",
                "type": "color",
                "group": "brand",
                "value": "#3f55b1",
                "slug": "brand_tertiary"
              }
            ]
          }
        ]
      },
      {
        "name": "Font Size",
        "groups": [
          {
            "name": "Content",
            "styles": [
              {
                "type": "font-size",
                "group": "content",
                "name": "Normal",
                "value": "0.9rem",
                "slug": "content_normal"
              }
            ]
          }
        ]
      }
    ]
  },
  {
    "name": "StyleGuide 3",
    "slug": "styleGuide3",
    "baseFontSize": 10,
    "types": [
      {
        "name": "Color",
        "groups": [
          {
            "name": "Base",
            "styles": [
              {
                "name": "Primary",
                "type": "color",
                "group": "base",
                "value": "#2be053",
                "slug": "base_primary"
              }
            ]
          }
        ]
      }
    ]
  }
]
