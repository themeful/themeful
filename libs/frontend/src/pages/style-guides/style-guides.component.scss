:host {
  display: block;
}

.style-guide {
  text-align: center;

  h3 {
    display: flex;
    justify-content: center;
    gap: 4px;
  }

  &__wrapper {
    display: flex;
    flex-direction: row;
    gap: 16px;
  }
  &__type {
    text-align: center;

    h4 {
      font: var(--tf-default-font);
      margin: 0;
      padding: 8px;
    }
  }
  &__group {
    text-align: center;
    border: 1px solid currentColor;
    border-radius: 8px;
    overflow: hidden;
    margin-bottom: 16px;

    h5 {
      font: var(--tf-default-font);
      border-bottom: 1px solid currentColor;
      margin: 0;
      padding: 8px;
    }

    tf-property:not(:last-of-type) {
      box-shadow: 0 1px 0 0 rgba(200, 200, 200, 0.55);
    }
  }
}
