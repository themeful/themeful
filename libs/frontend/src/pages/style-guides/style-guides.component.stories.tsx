import { h } from '@stencil/core'

export default {
  title: 'Pages/Style Guides',
  args: {
    first: 'John',
    middle: 'S',
    last: 'Doe',
  },
}

export const styleGuides = (args) => {
  return <tf-style-guides {...args}></tf-style-guides>
}
