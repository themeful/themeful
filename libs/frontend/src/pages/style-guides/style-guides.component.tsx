import { Component, h, Prop, State } from '@stencil/core'
import {
  ExtendedStyle,
  ExtendedStyleGuide,
  ExtendedStyleGuides,
  StyleGroup,
  StyleTypeGroup,
} from '@typings'
import { Observable, Subscription } from 'rxjs'

@Component({
  tag: 'tf-style-guides',
  styleUrl: 'style-guides.component.scss',
  shadow: true,
})
export class StyleGuidesComponent {
  /** Style Guides*/
  @Prop() styleGuides$: Observable<ExtendedStyleGuides>

  @State() sub?: Subscription
  @State() styleGuides: ExtendedStyleGuides

  componentWillLoad() {
    this.sub = this.styleGuides$?.subscribe((styleGuides) => {
      this.styleGuides = styleGuides
    })
  }

  disconnectedCallback() {
    this.sub?.unsubscribe()
  }

  public render(): HTMLTfStyleGuidesElement {
    return (
      <main class="style-guide__wrapper">
        {this.styleGuides &&
          this.styleGuides.map((styleGuide) => this.renderStyleGuide(styleGuide))}
      </main>
    )
  }

  private renderStyleGuide(styleGuide: ExtendedStyleGuide): HTMLElement {
    return (
      <div class="style-guide">
        <h3>{styleGuide.name}</h3>
        {styleGuide.types && styleGuide.types.map((type) => this.renderType(type, styleGuide.slug))}
      </div>
    )
  }

  private renderType(type: StyleTypeGroup, styleGuideSlug: string): HTMLElement {
    return (
      <div class="style-guide__type">
        <h4>{type.name}</h4>
        {type.groups &&
          type.groups.map((group) =>
            this.renderGroup(group, styleGuideSlug, type.groups.length > 1)
          )}
      </div>
    )
  }

  private renderGroup(group: StyleGroup, styleGuideSlug: string, showGroup: boolean): HTMLElement {
    return (
      <div class="style-guide__group property--grouped">
        {showGroup && <h5>{group.name}</h5>}
        {group.styles && group.styles.map((style) => this.renderStyle(style, styleGuideSlug))}
      </div>
    )
  }

  private renderStyle(property: ExtendedStyle, styleGuide: string): HTMLTfPropertyElement {
    return <tf-property {...{ property, styleGuide, showGroup: false }} />
  }
}
