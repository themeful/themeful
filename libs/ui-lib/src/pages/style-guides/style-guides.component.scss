:host {
  display: block;
  width: 100%;
  height: 100%;
}

.style-guide {
  text-align: center;
  position: relative;

  h3 {
    display: flex;
    justify-content: center;
    gap: 4px;
    max-width: 200px;
    height: 42px;
    position: sticky;
    top: 0;
    z-index: 1;
    background-color: var(--tf-content-bg-color);
    padding: 16px 8px;
    margin: 0;
  }

  &__wrapper {
    display: flex;
    flex-direction: row;
    // gap: 16px;
    padding: 0 16px;
    width: 100%;
    height: 100%;
    overflow: auto;
  }
  &__type {
    text-align: center;
    padding: 0 8px;

    h4 {
      font: var(--tf-default-font);
      margin: 0;
      padding: 8px;
    }
  }
  &__group {
    text-align: center;
    border: 1px solid currentColor;
    border-radius: 8px;
    overflow: hidden;
    margin-bottom: 16px;

    h5 {
      font: var(--tf-default-font);
      border-bottom: 1px solid currentColor;
      margin: 0;
      padding: 8px;
    }

    tf-property:not(:last-of-type) {
      box-shadow: 0 1px 0 0 rgba(200, 200, 200, 0.55);
    }
  }
}

a,
a:focus,
a:active {
  text-decoration: none;
  color: inherit;
}
