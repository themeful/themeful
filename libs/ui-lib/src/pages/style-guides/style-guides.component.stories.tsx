import { h } from '@stencil/core'
import { action } from '@storybook/addon-actions'
import { of } from 'rxjs'
import sample from './style-guides.sample.json'

export default {
  title: 'Pages/Style Guides',
  args: {
    styleGuides$: of(sample),
    onAction: action('triggered action'),
  },
}

export const styleGuides = (args) => {
  return (
    <div>
      <div class="header" style={{ marginBottom: '16px' }}>
        <h1>Style Guides</h1>
      </div>
      <div class="content">
        <div style={{ position: 'relative', height: '90vh' }}>
          <tf-style-guides {...args} />
        </div>
      </div>
    </div>
  )
}
