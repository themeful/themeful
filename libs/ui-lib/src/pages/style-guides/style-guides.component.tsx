import { propertySelect } from '@properties'
import { Component, Event, EventEmitter, h, Host, Prop, State } from '@stencil/core'
import {
  ExtendedStyle,
  ExtendedStyleGuide,
  ExtendedStyleGuides,
  FormIntegrationActions,
  StyleGroup,
  StyleGuideBase,
  StyleTypeGroup,
} from '@typings'
import { Observable, Subject, Subscription } from 'rxjs'
import { createRouter, href } from 'stencil-router-v2'
import '../../components/icon'
import '../../components/menu'
import '../../components/navigation'
import '../../components/property'
import '../../forms/form-integration'

const Router = createRouter()

@Component({
  tag: 'tf-style-guides',
  styleUrl: 'style-guides.component.scss',
  shadow: true,
})
export class StyleGuidesComponent {
  /** Style Guides */
  @Prop() styleGuides$!: Observable<ExtendedStyleGuides>

  @State() styleGuides!: ExtendedStyleGuides
  private formData$ = new Subject()

  /** Event emitted when an action is triggered */
  @Event({ composed: false }) action!: EventEmitter<FormIntegrationActions>

  private onAction = ({ detail }: { detail: FormIntegrationActions }): void => {
    if (detail.action !== 'close') {
      this.action.emit(detail)
    }
  }

  private sub = new Subscription()
  private styleGuideGroups!: { [styleguide: string]: string[] }

  private openStyleGuideForm = (styleGuide?: string, styleGuideBase?: StyleGuideBase): void => {
    this.formData$.next({
      form: 'styleguide',
      identifier: styleGuide,
      fields: styleGuideBase,
    })
  }

  private openStyleForm = (styleGuide: string, extendedStyle?: ExtendedStyle): void => {
    this.formData$.next({
      form: 'style',
      identifier: { styleGuide, style: extendedStyle?.slug },
      groups: this.styleGuideGroups[styleGuide],
      propertyTypes: propertySelect,
      fields: extendedStyle,
    })
  }

  private openStyleGuideDuplicateForm = (styleGuide: string, name: string): void => {
    this.formData$.next({
      form: 'styleguideDuplicate',
      identifier: styleGuide,
      fields: { name },
    })
  }

  public componentWillLoad(): void {
    this.sub.add(
      this.styleGuides$?.subscribe((styleGuides) => {
        this.styleGuides = styleGuides
        this.styleGuideGroups = styleGuides.reduce(
          (result: { [key: string]: string[] }, styleguide) => {
            result[styleguide.slug] = []
            styleguide.types.forEach((type) => {
              type.groups.forEach((group) => {
                result[styleguide.slug].push(group.name)
              })
            })
            return result
          },
          {}
        )
      })
    )
  }

  public disconnectedCallback(): void {
    this.sub.unsubscribe()
  }

  public render(): HTMLTfStyleGuidesElement {
    return (
      <Host>
        <stencil-route-title pageTitle="Themeful - Style Guides" />
        <tf-navigation
          items={[
            {
              label: 'Add Style Guide',
              callback: this.openStyleGuideForm,
            },
          ]}
          size="small"
        />
        <div class="style-guide__wrapper">
          {this.styleGuides &&
            this.styleGuides.map((styleGuide) => this.renderStyleGuide(styleGuide))}
        </div>
        <tf-form-integration {...{ formData$: this.formData$, onAction: this.onAction }} />
      </Host>
    )
  }

  private renderStyleGuide(styleGuide: ExtendedStyleGuide): HTMLElement {
    return (
      <div class="style-guide">
        <h3>
          <a {...href(`/styleguide/${styleGuide.slug}`)}>{styleGuide.name}</a>
          <tf-menu
            items={[
              {
                label: 'Edit',
                icon: 'pen',
                callback: () =>
                  this.openStyleGuideForm(styleGuide.slug, {
                    name: styleGuide.name,
                    baseFontSize: styleGuide.baseFontSize,
                  }),
              },
              {
                label: 'Add Style',
                icon: 'plus',
                callback: () => this.openStyleForm(styleGuide.slug),
              },
              {
                label: 'Show',
                icon: 'search',
                callback: () => {
                  Router.push(`/styleguide/${styleGuide.slug}`)
                },
              },
              {
                label: 'Duplicate',
                icon: 'copy',
                callback: () => {
                  this.openStyleGuideDuplicateForm(styleGuide.slug, styleGuide.name)
                },
              },
            ]}
          />
        </h3>
        {styleGuide.types && styleGuide.types.map((type) => this.renderType(type, styleGuide.slug))}
      </div>
    )
  }

  private renderType(type: StyleTypeGroup, styleGuideSlug: string): HTMLElement {
    return (
      <div class="style-guide__type">
        <h4>{type.name}</h4>
        {type.groups &&
          type.groups.map((group) =>
            this.renderGroup(group, styleGuideSlug, type.groups.length > 1)
          )}
      </div>
    )
  }

  private renderGroup(group: StyleGroup, styleGuideSlug: string, showGroup: boolean): HTMLElement {
    return (
      <div class="style-guide__group property--grouped">
        {showGroup && <h5>{group.name}</h5>}
        {group.styles && group.styles.map((style) => this.renderStyle(style, styleGuideSlug))}
      </div>
    )
  }

  private renderStyle(extendedStyle: ExtendedStyle, styleGuide: string): HTMLTfPropertyElement {
    return (
      <tf-property
        {...{
          extendedStyle,
          showGroup: false,
          onEdit: () => this.openStyleForm(styleGuide, extendedStyle),
        }}
      />
    )
  }
}
