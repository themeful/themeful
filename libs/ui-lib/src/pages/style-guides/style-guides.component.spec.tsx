import { newSpecPage } from '@stencil/core/testing'
import { StyleGuidesComponent } from './style-guides.component'

describe('StyleGuidesComponent', () => {
  it('renders', async () => {
    const { root } = await newSpecPage({
      components: [StyleGuidesComponent],
      html: '<tf-style-guides></tf-style-guides>',
      url: 'http://localhost/',
    })
    expect(root).toMatchSnapshot()
  })

  it('renders with values', async () => {
    const { root } = await newSpecPage({
      components: [StyleGuidesComponent],
      html: `<tf-style-guides></tf-style-guides>`,
      url: 'http://localhost/',
    })
    expect(root).toMatchSnapshot()
  })
})
