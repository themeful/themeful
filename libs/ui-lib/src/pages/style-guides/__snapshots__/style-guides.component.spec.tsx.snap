// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleGuidesComponent renders 1`] = `
<tf-style-guides>
  <mock:shadow-root>
    <stencil-route-title pagetitle="Themeful - Style Guides"></stencil-route-title>
    <tf-navigation size="small"></tf-navigation>
    <div class="style-guide__wrapper"></div>
    <tf-form-integration></tf-form-integration>
  </mock:shadow-root>
</tf-style-guides>
`;

exports[`StyleGuidesComponent renders with values 1`] = `
<tf-style-guides>
  <mock:shadow-root>
    <stencil-route-title pagetitle="Themeful - Style Guides"></stencil-route-title>
    <tf-navigation size="small"></tf-navigation>
    <div class="style-guide__wrapper"></div>
    <tf-form-integration></tf-form-integration>
  </mock:shadow-root>
</tf-style-guides>
`;
